### PR TITLE
Support for a remote parameter database

### PR DESCRIPTION
**Pull request by katzfey**
### Solved Problem
The voxl2 has a split architecture. PX4 runs on a posix platform and a Qurt platform. The two communicate uorb topics back and forth with the muorb module. But each has it's own parameters database and they need to stay in sync with each other. This PR adds support to keep the 2 parameter databases in sync. The main parameters database running on Linux has file system support while the Qurt one does not. The Linux side is considered the primary and the Qurt side is considered the remote.

### CODE REVIEW

File: src/lib/parameters/parameters.cpp [hunk 1]
```diff
@@ -116,6 +118,10 @@ static perf_counter_t param_set_perf;
 static pthread_mutex_t file_mutex  =
 	PTHREAD_MUTEX_INITIALIZER; ///< this protects against concurrent param saves (file or flash access).
 
+// Support for remote parameter node
+#include "parameters_primary.h"
```
Comment (dagar) [inline src/lib/parameters/parameters.cpp:122]:
```suggestion
#if defined(CONFIG_PARAM_PRIMARY)
# include "parameters_primary.h"
#endif // CONFIG_PARAM_PRIMARY
```

File: src/lib/parameters/parameters.cpp [hunk 2]
```diff
@@ -116,6 +118,10 @@ static perf_counter_t param_set_perf;
...
+#include "parameters_remote.h"
```
Comment (dagar) [inline src/lib/parameters/parameters.cpp:123]:
```suggestion
#if defined(CONFIG_PARAM_REMOTE)
# include "parameters_remote.h"
#endif // CONFIG_PARAM_REMOTE
```

File: src/lib/parameters/parameters.cpp [hunk 3]
```diff
@@ -128,14 +134,27 @@ param_init()
 	px4_register_boardct_ioctl(_PARAMIOCBASE, param_ioctl);
 #endif
 
+#if defined(CONFIG_PARAM_PRIMARY)
+	param_primary_init();
+#endif
```
Comment (dagar) [inline src/lib/parameters/parameters.cpp:139]:
```suggestion
#endif // CONFIG_PARAM_PRIMARY
```

File: src/lib/parameters/parameters.cpp [hunk 4]
```diff
@@ -128,14 +134,27 @@ param_init()
...
+
+#if defined(CONFIG_PARAM_REMOTE)
+	param_remote_init();
+#endif
```
Comment (dagar) [inline src/lib/parameters/parameters.cpp:143]:
```suggestion
#endif // CONFIG_PARAM_REMOTE
```

File: src/lib/parameters/parameters.cpp [hunk 5]
```diff
@@ -468,6 +512,14 @@ bool param_used(param_t param)
 void param_set_used(param_t param)
 {
 	if (handle_in_range(param)) {
+#if defined(CONFIG_PARAM_REMOTE)
+
+		if (! param_used(param)) {
```
Comment (dagar) [inline src/lib/parameters/parameters.cpp:517]:
```suggestion
		if (!param_used(param)) {
```

**Review by dagar — APPROVED** [review]
I think we can live with this for now as long as we're okay dealing with some churn down the line in order to get to a clean parameter server.

**Comment by mrpollo** [comment]
Happy to see this come in, good work @katzfey 👏

**Pull request merged**